arch/riscv/core/isr.S

```diff
@@ -758,31 +814,40 @@ on_thread_stack:
 
 #ifdef CONFIG_STACK_SENTINEL
 	call z_check_stack_sentinel
-	GET_CPU(t1, t2)
 #endif
 
-#ifdef CONFIG_PREEMPT_ENABLED
-	/*
-	 * Check if we need to perform a reschedule
-	 */
+reschedule:
+	/* Get next thread to run and switch to it - or not, if the same */
+	GET_CPU(t0, t1)
+	RV_OP_LOADREG t1, ___cpu_t_current_OFFSET(t0)
+	mv a0, t1
+
+	addi sp, sp, -16
+	RV_OP_STOREREG t1, 0x00(sp)
 
-	/* Get pointer to _current_cpu.current */
-	RV_OP_LOADREG t2, ___cpu_t_current_OFFSET(t1)
+	call z_get_next_switch_handle
```

Comment:
I don't know enough about what's happening here (or about RISC-V machine language frankly) to know for sure just from reading a diff, but I think this might be falling into the same trap that just bit arm64:  the switch handle argument passed to z_get_next_switch_handle() has to be a **completely saved** context.  When in SMP contexts, the scheduler will save that into global state and another CPU is free to pick it up and try to run it instantly.  If there are bits you're still finishing up, it may not see them in time.

The workaround on arm64 was to pass a null argument and then write the switch_handle field of the thread later (the null is used as a signal on which the other CPU can spin).

Reply:
That's good information! Indeed, userspace and some stack guard related code is still run after this point, so those may be in danger (if not userspace, everything is already saved at this point). And I guess there's no test for this trap (so it will be a rare kind of issue). 
Semi-unrelated - may I rant that a function named "get" does change the overall state? I spent some time debugging an issue that was related to this function changing `_current`. And I just noticed it already does `z_thread_mark_switched_out`.

Reply:
Yes.  z_get_next_switch_handle() started as a simple utility but evolved into the standardized "end of interrupt" hook for the kernel.  And yes, the name isn't great.  But as you might expect, changing the name of a function called from assembly code on every architecture is a high friction operation. :)  But yes: "z_interrupt_complete()" might be a better name.